### PR TITLE
Consolidate image reference generation/lookup

### PR DESCRIPTION
**Pull request by ncdc**
Add imageapi.LatestTaggedImage(repo, tag) to retrieve the most recent image
data (image id and pull spec) for a repo and tag.

Update controllers and generators to use this new helper.

### CODE REVIEW

**Comment by ncdc** [comment]
[test]

File: pkg/image/api/helper.go [hunk 1]
```diff
@@ -110,3 +115,22 @@ func ImageWithMetadata(image Image) (*Image, error) {
 
 	return &image, nil
 }
+
+// LatestTagEvent returns the most recent TagEvent for the specified image
+// repository and tag.
+func LatestTagEvent(repo ImageRepository, tag string) (*TagEvent, error) {
```
Comment (smarterclayton) [inline pkg/image/api/helper.go]: LatestTaggedImage

**Comment by ncdc** [comment]
@smarterclayton let me know if this looks ok or you need other changes

File: pkg/build/util/generate.go [hunk 2]
```diff
@@ -102,12 +102,13 @@ func GenerateBuildWithImageTag(config *buildapi.BuildConfig, revision *buildapi.
 		if len(tag) == 0 {
 			tag = buildapi.DefaultImageTag
 		}
-		imageID, hasTag := imageRepo.Tags[tag]
-		if !hasTag {
+		tagEvent, err := imageapi.LatestTaggedImage(*imageRepo, tag)
```
Comment (smarterclayton) [inline pkg/build/util/generate.go]: call tagEvent taggedImage

Comment (smarterclayton) [inline pkg/build/util/generate.go]: or just tag or `latest`

File: pkg/image/api/helper.go [hunk 3]
```diff
@@ -54,20 +54,25 @@ func IsPullSpec(spec string) bool {
 // JoinDockerPullSpec turns a set of components of a Docker pull specification into a single
 // string. Attempts to match as closely as possible the Docker spec up to 1.3. Future API
 // revisions may change the pull syntax.
-func JoinDockerPullSpec(registry, namespace, name, tag string) string {
-	if len(tag) != 0 {
-		tag = ":" + tag
+func JoinDockerPullSpec(registry, namespace, name, ref string) string {
+	if len(ref) != 0 {
```
Comment (smarterclayton) [inline pkg/image/api/helper.go]: I'm starting to think we need to refactor this into a struct - `PullSpec` with a set of fields, and digest and tag should not be the same.

Comment (smarterclayton) [inline pkg/image/api/helper.go]: I don't like inferring type based on : vs something else.

**Comment by ncdc** [comment]
@smarterclayton PTAL

**Comment by smarterclayton** [comment]
LGTM [merge]

**Pull request merged**